src/devices/src/display/gtk/worker.rs

```diff
@@ -0,0 +1,278 @@
+                                // ever anyway.
+                                unsafe { libc::_exit(125) };
+                            }
+                        });
+                    }
+                ))
+                .build(),
+            ActionEntry::builder("fullscreen")
+                .activate(glib::clone!(
+                    #[weak]
+                    window,
+                    move |_, _, _| {
+                        window.fullscreen();
+                    }
+                ))
+                .build(),
+            ActionEntry::builder("unfullscreen")
+                .activate(glib::clone!(
+                    #[weak]
+                    window,
+                    move |_, _, _| {
+                        window.unfullscreen();
+                    }
+                ))
+                .build(),
+        ]);
+        window.insert_action_group("scanout", Some(&actions));
+
+        let fullscreen_btn = gtk4::Button::builder()
+            .icon_name("view-fullscreen")
+            .tooltip_text("Enter fullscreen mode")
+            .action_name("scanout.fullscreen")
+            .build();
+
+        let picture = Picture::builder()
+            // Set picture dimension to the requested display size,
+            // this will make the created window have the appropriate size
+            .width_request(display_info.width as i32)
+            .height_request(display_info.height as i32)
+            .build();
+
+        window.set_titlebar(Some(&header_bar));
+        header_bar.pack_start(&build_kill_vm_btn());
+        header_bar.pack_end(&fullscreen_btn);
+
+        let overlay = build_overlay(&window);
+        overlay.set_child(Some(&picture));
+        window.set_child(Some(&overlay));
+        window.set_visible(true);
+
+        // Unset the width/height after the window is created to allow resizing the window to
+        // be smaller
+        picture.set_size_request(-1, -1);
+
+        Self {
+            window,
+            width,
+            height,
+            format,
+            picture,
+        }
+    }
+
+    fn update_params(&mut self, width: i32, height: i32, format: gdk::MemoryFormat) {
+        self.width = width;
+        self.height = height;
+        self.format = format;
+    }
+
+    fn update(&mut self, data: Vec<u8>, stride: u32) {
+        let texture = gdk::MemoryTexture::new(
+            self.width,
+            self.height,
+            self.format,
+            &glib::Bytes::from_owned(data),
+            stride as usize,
+        );
+        self.picture.set_paintable(Some(&texture));
+    }
+}
+
+impl Drop for Scanout {
+    fn drop(&mut self) {
+        self.window.destroy();
+    }
+}
+
+fn build_kill_vm_btn() -> Button {
+    let kill_vm_btn = Button::builder()
+        .label("Kill VM")
+        .action_name("scanout.kill")
+        .build();
+    kill_vm_btn.set_css_classes(&["destructive-action"]);
+    kill_vm_btn
+}
```

Comment:
seems too much to do this in a separate function. Also use `add_css_class`, the function you are using removes all the pre-existing CSS classes in the widget if any.